planner/funcdep/fd_graph.go

```diff
@@ -549,6 +549,280 @@ func (s *FDSet) MakeCartesianProduct(rhs *FDSet) {
 	// maintain a key?
 }
 
+// MakeOuterJoin generates the records the fdSet of the outer join.
+//
+// We always take the left as the row-supplying side, and right side as the null-supplying side. (swap it if not)
+// As we know, the outer join would generate null extended rows compared with inner join.
+// So we cannot directly do the same thing with the inner join. This function deals with the special cases of the outer join.
+//
+// Knowledge:
+// 1: the filter condition related to lhs column won't filter predicate-allowed rows and refuse null rows (left rows are always exist for all)
+// 2: the filter condition related to rhs column won't filter NULL rows although filter has not null attribute on it. (null-appending happened after that)
+//
+// Notification:
+// 1: the origin FD from left side (rows-supplying) over the result of outer join filtered are preserved, because
+//		it may be duplicated by multi matching, but actually they are same left rows (don't violate FD definition).
+//
+// 2: the origin FD from right side (nulls-supplying) over the result of outer join filtered may not be valid anymore.
+//
+//		<1> strict FD may be wakened as a lax one. But if at least one non-NULL column is part of the determinant, the
+//			strict FD can be preserved.
+//	        a  b  |  c     d     e
+//			------+----------------
+//		 	1  1  |  1    NULL   1
+//		    1  2  | NULL  NULL  NULL
+//		    2  1  | NULL  NULL  NULL
+//			left join with (a,b) * (c,d,e) on (a=c and b=1), if there is a strict FD {d} -> {e} on the rhs. After supplied
+//			with null values, {d} -> {e} are degraded to a lax one {d} ~~> {e} as you see. the origin and supplied null value
+//			for d column determine different dependency.  NULL -> 1 and NULL -> NULL which breaks strict FD definition.
+//
+//			Unless the determinant contains at least a not null column for example c here, FD like {c,d} -> {e} can survive
+//			after the left join. Because you can not find two same key, one from the origin rows and the other one from the
+//			supplied rows.
+//
+//			for lax FD, the supplied rows of null values doesn't take effect of lax FD itself. So it can be kept.
+//
+//		<2> constant FD should be removed, since null values may be substituted for some unmatched left rows. NULL is not a
+//			constant anymore.
+//
+//      <3> equivalence FD should be removed, since substituted null value are not equal to the other substituted null value.
+//
+// 3: the newly added FD from filters, should take some considerations as below:
+//
+//	 	<1> strict/lax FD: join key filter conditions can not produce new strict/lax FD yet (knowledge: 1&2).
```

Comment:
What's the first place?

Reply:
I mean, we can't derive strict/lax FD from a filter in fact.

Reply:
Yes, for a more general situation.
We are generating the funcdep from:
- Projection
- Index constraint of the table.
- The aggregation.

For case#2, we're curious about whether the from side(the keys) can be null or not. For case #3, we know whether the group-by items can be null or not.

For case #1, we can regard the project as `(x, y, ...z) to f(x, y, ...z)`. We are easy to know whether the `(x, y, ...z)` is null or not, but not easy to the `f` is null or not. 

So we only consider the from side.

Reply:
It cannot directly produce any funcdep just the same as the case when it serve as a normal filter of inner join.
It can produce funcdep with more strict constraints.

Reply:
actually. it‘s possible.   
select * from t left join t1 on (case when (t.a>1) then (t1.b=3) else (t1.b=2) end);
the join predicate will derive strict FD {a} -> {b} even supplied with null value.

Exploring this kind of predicate is out of reach for now, we can dig more detail in the short future